File: typescript/packages/core/src/evaluator.ts
/**
 * Policy evaluation engine - deterministic rule evaluation.
 */

import { randomUUID } from 'node:crypto';
import {
  EnvelopeSchema,
  PolicySetSchema,
  DecisionType,
  type Envelope,
  type PolicySet,
  type Decision,
  type Rule,
  type Condition,
  type LogicalCondition,
  type MatchedRule,
} from './types.js';
import { resolveFieldPath } from './utils.js';
import { evaluateOperator } from './operators.js';

/**
 * Policy evaluator - deterministic rule evaluation engine.
 *
 * This is the core runtime enforcement engine. It evaluates policies (JSON objects)
 * against execution envelopes and returns decisions.
 *
 * **Design principles:**
 * - Deterministic: same (envelope, policy) always produces same decision
 * - Side-effect free: evaluation does not modify state
 * - Fast: sub-millisecond evaluation for typical policies
 * - Safe-by-default: no matching rule → BLOCK
 *
 * **Evaluation semantics:**
 * - Rules evaluated by priority (descending order)
 * - First-match wins (returns decision from first matching rule)
 * - Disabled rules are skipped
 * - Logical operators: all_of (AND), any_of (OR), none_of (NOT)
 *
 * @example
 * ```typescript
 * const evaluator = new PolicyEvaluator();
 * const decision = evaluator.evaluate(envelope, policy);
 * if (decision.allowed) {
 *   await executeAction();
 * }
 * ```
 */
export class PolicyEvaluator {
  /**
   * Evaluate a policy against an execution envelope.
   *
   * This is the main entry point for policy evaluation. It validates inputs,
   * evaluates rules, and returns a decision response with timing metadata.
   *
   * **Timing breakdown:**
   * - `ingest_ms`: Input validation time (Zod schema parsing)
   * - `evaluation_ms`: Rule evaluation time
   * - `total_ms`: End-to-end time (ingest + evaluation)
   *
   * @param envelope - Execution envelope (validated against envelope schema)
   * @param policy - Policy set (validated against policy schema)
   * @returns Decision response with action decision, reason codes, and timing
   *
   * @throws {ZodError} If envelope or policy validation fails
   *
   * @example
   * ```typescript
   * const decision = evaluator.evaluate(
   *   { action_id: 'act_123', org_id: 'org_abc', ... },
   *   { name: 'payment-controls', version: '1.0', rules: [...] }
   * );
   * console.log(decision.decision);  // 'ALLOW' | 'BLOCK' | 'PAUSE' | ...
   * console.log(decision.allowed);   // true | false
   * console.log(decision.timing.total_ms);  // e.g., 0.42
   * ```
   */
  evaluate(
    envelope: Envelope | Record<string, unknown>,
    policy: PolicySet | Record<string, unknown>,
    mode: 'OBSERVE_ALL' | 'RESPECT_POLICY' = 'RESPECT_POLICY'
  ): Decision {
    const startTime = performance.now();

    // Step 1: Validate inputs (Zod schema parsing)
    const validatedEnvelope = EnvelopeSchema.parse(envelope);
    const validatedPolicy = PolicySetSchema.parse(policy);
    const ingestMs = performance.now() - startTime;

    // Step 2: Evaluate rules
    const evalStart = performance.now();
    const [decisionType, reasonCodes, matchedRules, matchedRuleObj] = this.evaluateRules(
      validatedEnvelope,
      validatedPolicy
    );
    const evaluationMs = performance.now() - evalStart;

    // Step 3: Build decision response
    const totalMs = performance.now() - startTime;

    // Map decision type to allowed flag
    // ALLOW and SANDBOX are the only "allowed" decisions
    const allowed = decisionType === DecisionType.ALLOW || decisionType === DecisionType.SANDBOX;

    // Remediation: pass through from matched rule (only for blocking decisions)
    const remediation =
      !allowed && matchedRuleObj?.remediation ? matchedRuleObj.remediation : undefined;

    // Check if this decision should be wrapped in OBSERVE mode
    let shouldObserve = false;
    if (mode === 'OBSERVE_ALL' && !allowed) {
      shouldObserve = true;
    } else if (
      mode === 'RESPECT_POLICY' &&
      matchedRuleObj &&
      (matchedRuleObj.mode ?? 'enforce') === 'observe' &&
      !allowed
    ) {
      shouldObserve = true;
    }

    if (shouldObserve) {
      const observeTotalMs = performance.now() - startTime;
      return {
        action_id: validatedEnvelope.action_id,
        decision: DecisionType.OBSERVE,
        allowed: true,
        reason_codes: ['OBSERVED'],
        would_be: decisionType,
        would_be_reason_codes: reasonCodes,
        policy_version: validatedPolicy.version,
        timing: {
          ingest_ms: Number(ingestMs.toFixed(3)),
          evaluation_ms: Number(evaluationMs.toFixed(3)),
          total_ms: Number(observeTotalMs.toFixed(3)),
        },
        matched_rules: matchedRules.length > 0 ? matchedRules : undefined,
      };
    }

    // Escalation fields: resume_token + route_ref
    const escalationTypes: Set<string> = new Set([DecisionType.REQUIRE_APPROVAL, DecisionType.PAUSE, DecisionType.ESCALATE]);
    let resumeToken: string | undefined;
    let routeRef: string | undefined;
    if (escalationTypes.has(decisionType)) {
      resumeToken = `rtk_${randomUUID().replace(/-/g, '')}`;
      if (matchedRuleObj?.route) {
        routeRef = matchedRuleObj.route;
      }
    }

    return {
      action_id: validatedEnvelope.action_id,
      decision: decisionType,
      allowed,
      reason_codes: reasonCodes,
      policy_version: validatedPolicy.version,
      timing: {
        ingest_ms: Number(ingestMs.toFixed(3)),
        evaluation_ms: Number(evaluationMs.toFixed(3)),
        total_ms: Number(totalMs.toFixed(3)),
      },
      matched_rules: matchedRules.length > 0 ? matchedRules : undefined,
      resume_token: resumeToken,
      route_ref: routeRef,
      remediation,
    };
  }

  /**
   * Evaluate all rules in a policy set.
   *
   * **Evaluation semantics:**
   * 1. Sort rules by priority (descending - highest priority first)
   * 2. Iterate through rules in order
   * 3. Skip disabled rules
   * 4. Return decision from first matching rule (first-match wins)
   * 5. If no rule matches, return BLOCK (safe-by-default)
   *
   * @param envelope - Validated execution envelope
   * @param policy - Validated policy set
   * @returns Tuple of [decision type, reason codes, matched rules]
   *
   * @private
   */
  private evaluateRules(
    envelope: Envelope,
    policy: PolicySet
  ): [DecisionType, string[], MatchedRule[], Rule | undefined] {
    // Step 1: Sort rules by priority (descending order)
    // Important: Create a copy to avoid mutating the original policy
    const sortedRules = [...policy.rules].sort((a, b) => b.priority - a.priority);

    // Step 2: First-match semantics - return on first matching rule
    for (const rule of sortedRules) {
      // Skip disabled rules
      if (!rule.enabled) {
        continue;
      }

      // Evaluate rule conditions
      if (this.evaluateCondition(envelope, rule.conditions)) {
        // Rule matched - return its decision
        const matchedRule: MatchedRule = {
          rule_name: rule.name,
          policy_set: policy.name,
          priority: rule.priority,
        };

        return [rule.decision, [rule.reason_code], [matchedRule], rule];
      }
    }

    // Step 3: Safe-by-default - no matching rule → BLOCK
    return [DecisionType.BLOCK, ['NO_MATCHING_RULE'], [], undefined];
  }

  /**
   * Evaluate a condition (atomic or logical).
   *
   * This method discriminates between atomic and logical conditions
   * using type discrimination on the 'field' property.
   *
   * @param envelope - Validated execution envelope
   * @param condition - Condition to evaluate (atomic or logical)
   * @returns True if condition matches, false otherwise
   *
   * @private
   */
  private evaluateCondition(envelope: Envelope, condition: Condition | LogicalCondition): boolean {
    // Type discrimination: check if it's an atomic condition (has 'field' property)
    if ('field' in condition) {
      return this.evaluateAtomicCondition(envelope, condition);
    } else {
      return this.evaluateLogicalCondition(envelope, condition);
    }
  }

  /**
   * Evaluate a logical condition (all_of, any_of, none_of).
   *
   * **Logical operators:**
   * - `all_of` (AND): All nested conditions must be true
   * - `any_of` (OR): At least one nested condition must be true
   * - `none_of` (NOT): None of the nested conditions may be true
   *
   * Uses short-circuit evaluation for performance:
   * - `all_of`: Stops on first false
   * - `any_of`: Stops on first true
   * - `none_of`: Stops on first true (then inverts)
   *
   * @param envelope - Validated execution envelope
   * @param condition - Logical condition
   * @returns True if logical condition matches, false otherwise
   *
   * @private
   */
  private evaluateLogicalCondition(envelope: Envelope, condition: LogicalCondition): boolean {
    if (condition.all_of) {
      // AND: all conditions must be true (short-circuit on first false)
      return condition.all_of.every((c) => this.evaluateCondition(envelope, c));
    } else if (condition.any_of) {
      // OR: at least one must be true (short-circuit on first true)
      return condition.any_of.some((c) => this.evaluateCondition(envelope, c));
    } else if (condition.none_of) {
      // NOT: none may be true (short-circuit on first true, then invert)
      return !condition.none_of.some((c) => this.evaluateCondition(envelope, c));
    }

    // Should never happen due to Zod validation (exactly one operator must be set)
    throw new Error('Logical condition must have all_of, any_of, or none_of');
  }

  /**
   * Evaluate an atomic condition.
   *
   * **Steps:**
   * 1. Resolve field path from envelope (e.g., "parameters.amount")
   * 2. Evaluate operator against field value and compare value
   *
   * **Null handling:**
   * - If field path resolves to null/undefined, operator evaluation handles it
   * - Most operators return false for null (except 'exists')
   *
   * @param envelope - Validated execution envelope
   * @param condition - Atomic condition with field, operator, and value
   * @returns True if atomic condition matches, false otherwise
   *
   * @private
   */
  private evaluateAtomicCondition(
    envelope: Envelope,
    condition: Condition & { field: string }
  ): boolean {
    // Step 1: Resolve field path (e.g., "parameters.amount" → 1000)
    const fieldValue = resolveFieldPath(envelope, condition.field);

    // Step 2: Evaluate operator (handles null values internally)
    return evaluateOperator(fieldValue, condition.operator, condition.value);
  }
}

/**
 * Convenience function for one-off evaluations.
 *
 * Creates a new PolicyEvaluator instance and evaluates the policy.
 * For repeated evaluations, create a PolicyEvaluator instance and reuse it.
 *
 * @param envelope - Execution envelope
 * @param policy - Policy set
 * @returns Decision response
 *
 * @example
 * ```typescript
 * import { evaluate } from '@hiitl/core';
 *
 * const decision = evaluate(envelope, policy);
 * if (decision.allowed) {
 *   await executeAction();
 * }
 * ```
 */
export function evaluate(
  envelope: Envelope | Record<string, unknown>,
  policy: PolicySet | Record<string, unknown>,
  mode: 'OBSERVE_ALL' | 'RESPECT_POLICY' = 'RESPECT_POLICY'
): Decision {
  const evaluator = new PolicyEvaluator();
  return evaluator.evaluate(envelope, policy, mode);
}

File: python/hiitl/core/evaluator.py
"""Policy evaluator - deterministic rule evaluation engine.

This is the core runtime enforcement engine. It evaluates policies (JSON objects)
against execution envelopes and returns decisions.

Source of truth:
- specs/envelope_schema.json
- docs/specs/policy_format.md
- docs/specs/decision_response.md

Design principles:
- Deterministic: same (envelope, policy) always produces same decision
- Side-effect free: evaluation does not modify state
- Fast: sub-millisecond evaluation for typical policies
- Language-neutral: behavior validated by conformance test suite
"""

import re
import time
import uuid
from typing import Any, Dict, List, Optional, Union

from .types import (
    Condition,
    ConditionOperator,
    Decision,
    DecisionType,
    Envelope,
    LogicalCondition,
    MatchedRule,
    PolicySet,
    Rule,
    Timing,
)


class PolicyEvaluator:
    """Policy evaluator - evaluates policies against envelopes.

    This evaluator works on JSON policy objects (Python dicts/Pydantic models).
    It does not handle YAML parsing - that's a separate concern.

    Example:
        >>> evaluator = PolicyEvaluator()
        >>> policy = PolicySet(**policy_dict)  # From JSON
        >>> envelope = Envelope(**envelope_dict)  # From JSON
        >>> decision = evaluator.evaluate(envelope, policy)
    """

    def __init__(self):
        """Initialize the policy evaluator."""
        pass

    def evaluate(
        self,
        envelope: Union[Envelope, Dict[str, Any]],
        policy: Union[PolicySet, Dict[str, Any]],
        mode: str = "RESPECT_POLICY",
    ) -> Decision:
        """Evaluate an envelope against a policy and return a decision.

        Args:
            envelope: Execution envelope (Envelope model or dict)
            policy: Policy set (PolicySet model or dict)
            mode: Evaluation mode. "OBSERVE_ALL" wraps all blocking decisions
                in OBSERVE (allowed=True). "RESPECT_POLICY" respects per-rule
                mode field (default).

        Returns:
            Decision object with outcome, reason codes, and timing metadata

        Raises:
            ValidationError: If envelope or policy doesn't match schema
        """
        start_time = time.perf_counter()

        # Validate and convert to Pydantic models
        if not isinstance(envelope, Envelope):
            envelope = Envelope(**envelope)

        if not isinstance(policy, PolicySet):
            policy = PolicySet(**policy)

        ingest_end = time.perf_counter()
        ingest_ms = (ingest_end - start_time) * 1000

        # Evaluate policy rules
        eval_start = time.perf_counter()
        decision_type, reason_codes, matched_rules, matched_rule_obj = (
            self._evaluate_rules(envelope, policy)
        )
        eval_end = time.perf_counter()
        evaluation_ms = (eval_end - eval_start) * 1000

        # Build decision response
        total_ms = (eval_end - start_time) * 1000

        # Escalation fields: resume_token + route_ref
        escalation_types = {DecisionType.REQUIRE_APPROVAL, DecisionType.PAUSE, DecisionType.ESCALATE}
        resume_token = None
        route_ref = None
        if decision_type in escalation_types:
            resume_token = f"rtk_{uuid.uuid4().hex}"
            if matched_rule_obj and matched_rule_obj.route:
                route_ref = matched_rule_obj.route

        # Remediation: pass through from matched rule (only for blocking decisions)
        allowed = decision_type in [DecisionType.ALLOW, DecisionType.SANDBOX]
        remediation = None
        if not allowed and matched_rule_obj and matched_rule_obj.remediation:
            remediation = matched_rule_obj.remediation

        # Check if this decision should be wrapped in OBSERVE mode
        should_observe = False
        if mode == "OBSERVE_ALL" and not allowed:
            should_observe = True
        elif (mode == "RESPECT_POLICY" and matched_rule_obj is not None
              and matched_rule_obj.mode == "observe" and not allowed):
            should_observe = True

        if should_observe:
            return Decision(
                action_id=envelope.action_id,
                decision=DecisionType.OBSERVE,
                allowed=True,
                reason_codes=["OBSERVED"],
                would_be=decision_type,
                would_be_reason_codes=reason_codes,
                policy_version=policy.version,
                timing=Timing(
                    ingest_ms=ingest_ms,
                    evaluation_ms=evaluation_ms,
                    total_ms=total_ms,
                ),
                matched_rules=matched_rules if matched_rules else None,
            )

        decision = Decision(
            action_id=envelope.action_id,
            decision=decision_type,
            allowed=allowed,
            reason_codes=reason_codes,
            policy_version=policy.version,
            timing=Timing(
                ingest_ms=ingest_ms,
                evaluation_ms=evaluation_ms,
                total_ms=total_ms,
            ),
            matched_rules=matched_rules if matched_rules else None,
            resume_token=resume_token,
            route_ref=route_ref,
            remediation=remediation,
        )

        return decision

    def _evaluate_rules(
        self, envelope: Envelope, policy: PolicySet
    ) -> tuple[DecisionType, List[str], List[MatchedRule], Optional[Rule]]:
        """Evaluate rules in priority order and return first match.

        Args:
            envelope: Validated envelope
            policy: Validated policy set

        Returns:
            Tuple of (decision_type, reason_codes, matched_rules, matched_rule_obj)
        """
        # Sort rules by priority (highest first)
        sorted_rules = sorted(
            policy.rules, key=lambda r: r.priority, reverse=True
        )

        # Evaluate rules in priority order (first match wins)
        for rule in sorted_rules:
            # Skip disabled rules
            if not rule.enabled:
                continue

            # Evaluate rule conditions
            if self._evaluate_condition(envelope, rule.conditions):
                # Rule matched!
                matched_rule = MatchedRule(
                    rule_name=rule.name,
                    policy_set=policy.name,
                    priority=rule.priority,
                )
                return (
                    rule.decision,
                    [rule.reason_code],
                    [matched_rule],
                    rule,
                )

        # No rules matched - default to BLOCK (safe by default)
        return (DecisionType.BLOCK, ["NO_MATCHING_RULE"], [], None)

    def _evaluate_condition(
        self, envelope: Envelope, condition: Union[Condition, LogicalCondition]
    ) -> bool:
        """Evaluate a condition (atomic or logical) against an envelope.

        Args:
            envelope: Envelope to evaluate against
            condition: Condition to evaluate (atomic or logical)

        Returns:
            True if condition matches, False otherwise
        """
        if isinstance(condition, Condition):
            return self._evaluate_atomic_condition(envelope, condition)
        elif isinstance(condition, LogicalCondition):
            return self._evaluate_logical_condition(envelope, condition)
        else:
            raise ValueError(f"Unknown condition type: {type(condition)}")

    def _evaluate_logical_condition(
        self, envelope: Envelope, condition: LogicalCondition
    ) -> bool:
        """Evaluate a logical condition (all_of, any_of, none_of).

        Args:
            envelope: Envelope to evaluate against
            condition: Logical condition

        Returns:
            True if logical condition matches, False otherwise
        """
        if condition.all_of is not None:
            # AND: all conditions must be true
            return all(
                self._evaluate_condition(envelope, cond)
                for cond in condition.all_of
            )
        elif condition.any_of is not None:
            # OR: at least one condition must be true
            return any(
                self._evaluate_condition(envelope, cond)
                for cond in condition.any_of
            )
        elif condition.none_of is not None:
            # NOT: none of the conditions may be true
            return not any(
                self._evaluate_condition(envelope, cond)
                for cond in condition.none_of
            )
        else:
            raise ValueError("Logical condition must have one of: all_of, any_of, none_of")

    def _evaluate_atomic_condition(
        self, envelope: Envelope, condition: Condition
    ) -> bool:
        """Evaluate an atomic condition (field comparison).

        Args:
            envelope: Envelope to evaluate against
            condition: Atomic condition

        Returns:
            True if condition matches, False otherwise
        """
        # Resolve field path (supports dot notation)
        field_value = self._resolve_field_path(envelope, condition.field)

        # Evaluate operator
        return self._evaluate_operator(
            field_value, condition.operator, condition.value
        )

    def _resolve_field_path(self, envelope: Envelope, field_path: str) -> Any:
        """Resolve a field path (dot notation) in the envelope.

        Examples:
            tool_name -> envelope.tool_name
            parameters.amount -> envelope.parameters["amount"]
            target.account_id -> envelope.target["account_id"]

        Args:
            envelope: Envelope to resolve field in
            field_path: Dot-notation field path

        Returns:
            Field value, or None if field doesn't exist
        """
        parts = field_path.split(".")
        current = envelope

        for part in parts:
            # Backward compat: tool_name → action
            if part == "tool_name":
                part = "action"
            if isinstance(current, dict):
                current = current.get(part)
            elif hasattr(current, part):
                current = getattr(current, part)
            else:
                return None

            if current is None:
                return None

        return current

    def _evaluate_operator(
        self, field_value: Any, operator: ConditionOperator, compare_value: Any
    ) -> bool:
        """Evaluate a comparison operator.

        Args:
            field_value: Value from envelope
            operator: Comparison operator
            compare_value: Value to compare against

        Returns:
            True if comparison matches, False otherwise
        """
        # Handle None/null fields
        if operator == ConditionOperator.EXISTS:
            return (field_value is not None) == compare_value

        # If field doesn't exist and operator isn't EXISTS, return False
        if field_value is None:
            return False

        # Equality operators
        if operator == ConditionOperator.EQUALS:
            return field_value == compare_value
        elif operator == ConditionOperator.NOT_EQUALS:
            return field_value != compare_value

        # Numeric comparison operators
        elif operator == ConditionOperator.GREATER_THAN:
            return field_value > compare_value
        elif operator == ConditionOperator.GREATER_THAN_OR_EQUAL:
            return field_value >= compare_value
        elif operator == ConditionOperator.LESS_THAN:
            return field_value < compare_value
        elif operator == ConditionOperator.LESS_THAN_OR_EQUAL:
            return field_value <= compare_value

        # String/array operators
        elif operator == ConditionOperator.CONTAINS:
            if isinstance(field_value, str):
                return compare_value in field_value
            elif isinstance(field_value, list):
                return compare_value in field_value
            else:
                return False
        elif operator == ConditionOperator.NOT_CONTAINS:
            if isinstance(field_value, str):
                return compare_value not in field_value
            elif isinstance(field_value, list):
                return compare_value not in field_value
            else:
                return True

        elif operator == ConditionOperator.STARTS_WITH:
            return isinstance(field_value, str) and field_value.startswith(
                compare_value
            )
        elif operator == ConditionOperator.ENDS_WITH:
            return isinstance(field_value, str) and field_value.endswith(
                compare_value
            )
        elif operator == ConditionOperator.MATCHES:
            # Regex match (use sparingly - can be slow)
            if not isinstance(field_value, str):
                return False
            try:
                pattern = re.compile(compare_value)
                return pattern.search(field_value) is not None
            except re.error:
                return False

        # Set operators
        elif operator == ConditionOperator.IN:
            return field_value in compare_value
        elif operator == ConditionOperator.NOT_IN:
            return field_value not in compare_value

        else:
            raise ValueError(f"Unknown operator: {operator}")


# Convenience function
def evaluate(
    envelope: Union[Envelope, Dict[str, Any]],
    policy: Union[PolicySet, Dict[str, Any]],
    mode: str = "RESPECT_POLICY",
) -> Decision:
    """Evaluate an envelope against a policy (convenience function).

    Args:
        envelope: Execution envelope (Envelope model or dict)
        policy: Policy set (PolicySet model or dict)
        mode: Evaluation mode ("OBSERVE_ALL" or "RESPECT_POLICY")

    Returns:
        Decision object

    Example:
        >>> from hiitl.core import evaluate
        >>> decision = evaluate(envelope_dict, policy_dict)
        >>> if decision.allowed:
        >>>     execute_action()
    """
    evaluator = PolicyEvaluator()
    return evaluator.evaluate(envelope, policy, mode=mode)
